src/charm.py

```diff
@@ -252,6 +333,312 @@ def _on_nginx_route_broken(self, _: Any) -> None:
+            A list containing service and additional hostnames
+        """
+        # The relation will always exist when this method is called
+        relation = self._get_relation()  # type: ignore[arg-type]
+        definition = self._get_definition_from_relation(relation)  # type: ignore[arg-type]
+        hostnames = [definition.service_hostname]
+        hostnames.extend(definition.additional_hostnames)
+        return hostnames
+
+    def _on_certificates_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the TLS Certificate relation created event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
+            event.defer()
+            return
+        relation = self._get_relation()
+        if relation is None:
+            self._cleanup()
+            self.unit.status = WaitingStatus("waiting for relation")
+            event.defer()
+            return
+        hostnames = self.get_additional_hostnames()
+        for hostname in hostnames:
+            private_key_password = self._tls.generate_password().encode()
+            private_key = generate_private_key(password=private_key_password)
+            private_key_dict = {
+                "password": private_key_password.decode(),
+                "key": private_key.decode(),
+            }
+            if JujuVersion.from_environ().has_secrets:
+                try:
+                    secret = self.model.get_secret(label=f"private-key-{hostname}")
+                    secret.set_content(private_key_dict)
+                except SecretNotFoundError:
+                    secret = self.app.add_secret(
+                        content=private_key_dict, label=f"private-key-{hostname}"
+                    )
+            self._tls.update_relation_data_fields(
+                private_key_dict, tls_certificates_relation, self.app
+            )
+            peer_relation = self.model.get_relation("nginx-peers")
+            if peer_relation:
+                self._tls.update_relation_data_fields(private_key_dict, peer_relation, self.app)
+
+    def _on_certificates_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Handle the TLS Certificate relation joined event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
+            event.defer()
+            return
+        relation = self._get_relation()
+        if relation is None:
+            self._cleanup()
+            self.unit.status = WaitingStatus("waiting for relation")
+            return
```

Comment:
This precondition check is repeated multiple times, I think some factoring could help